File: frontend/src/lib/api-client.ts
import { clearStoredAuth, getStoredAuth, setStoredAuth } from '@/lib/auth-storage'
import type { AuthResponse } from '@/types/auth'
import type { ApiError } from '@/types/task'

const BASE_URL = import.meta.env.VITE_API_URL ?? 'http://localhost:8080/api'

export class ApiRequestError extends Error {
  status: number
  fields?: Record<string, string>

  constructor(apiError: ApiError) {
    super(apiError.message)
    this.status = apiError.status
    this.fields = apiError.fields
  }
}

// Concurrent 401s must share a single refresh call, not fire one each.
let refreshPromise: Promise<string | null> | null = null

async function refreshAccessToken(): Promise<string | null> {
  const stored = getStoredAuth()
  if (!stored) return null

  if (!refreshPromise) {
    refreshPromise = fetch(`${BASE_URL}/auth/refresh`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ refresh_token: stored.refreshToken }),
    })
      .then(async (response) => {
        if (!response.ok) return null
        const data = (await response.json()) as AuthResponse
        setStoredAuth({
          accessToken: data.access_token,
          refreshToken: data.refresh_token,
          user: data.user,
        })
        return data.access_token
      })
      .catch(() => null)
      .finally(() => {
        refreshPromise = null
      })
  }

  return refreshPromise
}

interface ApiFetchOptions {
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE'
  body?: unknown
  /** Public auth endpoints (register/login/refresh): no bearer header, no refresh-retry on 401. */
  public?: boolean
}

export async function apiFetch<T>(path: string, options: ApiFetchOptions = {}): Promise<T> {
  const { method = 'GET', body, public: isPublic = false } = options

  const request = (accessToken?: string) =>
    fetch(`${BASE_URL}${path}`, {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...(accessToken ? { Authorization: `Bearer ${accessToken}` } : {}),
      },
      body: body !== undefined ? JSON.stringify(body) : undefined,
    })

  const stored = getStoredAuth()
  const response = await request(isPublic ? undefined : stored?.accessToken)

  if (response.status === 401 && !isPublic && stored) {
    const newAccessToken = await refreshAccessToken()
    if (newAccessToken) {
      return handleResponse<T>(await request(newAccessToken))
    }
    clearStoredAuth()
    window.location.assign('/login')
    throw new ApiRequestError({
      status: 401,
      message: 'Session expired',
      timestamp: new Date().toISOString(),
      path,
    })
  }

  return handleResponse<T>(response)
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (response.status === 204) return undefined as T

  const isJson = response.headers.get('content-type')?.includes('application/json')
  const data = isJson ? await response.json() : undefined

  if (!response.ok) {
    throw new ApiRequestError(
      (data as ApiError) ?? {
        status: response.status,
        message: response.statusText || 'Request failed',
        timestamp: new Date().toISOString(),
        path: '',
      },
    )
  }

  return data as T
}

File: mobile/lib/core/api/api_client.dart
import 'package:dio/dio.dart';

import '../../models/auth_response.dart';
import '../storage/auth_storage.dart';
import 'api_exception.dart';

/// Thin wrapper around a Dio instance that attaches the bearer access token
/// to every non-public request and, on a 401, shares a single in-flight
/// refresh call across any concurrently-failing requests before retrying
/// once — porting `refreshAccessToken()` from the web app's
/// `frontend/src/lib/api-client.ts`.
class ApiClient {
  final Dio dio;
  final AuthStorage _authStorage;
  late final Dio _refreshDio;

  /// Called when a 401 survives a refresh attempt (refresh token itself is
  /// invalid/expired) — the app should drop back to the login screen.
  void Function()? onSessionExpired;

  Future<String?>? _refreshFuture;

  ApiClient({required String baseUrl, AuthStorage? authStorage})
    : _authStorage = authStorage ?? AuthStorage(),
      dio = Dio(BaseOptions(baseUrl: baseUrl, contentType: 'application/json')) {
    _refreshDio = Dio(BaseOptions(baseUrl: baseUrl, contentType: 'application/json'));
    dio.interceptors.add(
      InterceptorsWrapper(onRequest: _onRequest, onError: _onError),
    );
  }

  Future<void> _onRequest(
    RequestOptions options,
    RequestInterceptorHandler handler,
  ) async {
    if (options.extra['public'] != true) {
      final stored = await _authStorage.read();
      if (stored != null) {
        options.headers['Authorization'] = 'Bearer ${stored.accessToken}';
      }
    }
    handler.next(options);
  }

  Future<void> _onError(DioException error, ErrorInterceptorHandler handler) async {
    final isPublic = error.requestOptions.extra['public'] == true;
    final alreadyRetried = error.requestOptions.extra['retried'] == true;

    if (error.response?.statusCode == 401 && !isPublic && !alreadyRetried) {
      final newAccessToken = await _refreshAccessToken();
      if (newAccessToken != null) {
        final retryOptions = error.requestOptions
          ..extra['retried'] = true
          ..headers['Authorization'] = 'Bearer $newAccessToken';
        try {
          final response = await dio.fetch(retryOptions);
          return handler.resolve(response);
        } on DioException catch (retryError) {
          return handler.next(retryError);
        }
      }
      await _authStorage.clear();
      onSessionExpired?.call();
    }

    handler.next(error);
  }

  Future<String?> _refreshAccessToken() {
    return _refreshFuture ??= _doRefresh().whenComplete(() => _refreshFuture = null);
  }

  Future<String?> _doRefresh() async {
    final stored = await _authStorage.read();
    if (stored == null) return null;
    try {
      final response = await _refreshDio.post<Map<String, dynamic>>(
        '/auth/refresh',
        data: {'refresh_token': stored.refreshToken},
      );
      final auth = AuthResponse.fromJson(response.data!);
      await _authStorage.write(
        StoredAuth(
          accessToken: auth.accessToken,
          refreshToken: auth.refreshToken,
          user: auth.user,
        ),
      );
      return auth.accessToken;
    } catch (_) {
      return null;
    }
  }

  /// Runs [body] and converts any [DioException] into an [ApiException]
  /// built from the backend's error shape (docs/api.md), so callers only
  /// ever need to catch one exception type.
  Future<T> guard<T>(Future<T> Function() body) async {
    try {
      return await body();
    } on DioException catch (e) {
      final data = e.response?.data;
      if (data is Map<String, dynamic>) {
        throw ApiException.fromJson(data);
      }
      if (e.type == DioExceptionType.connectionError ||
          e.type == DioExceptionType.connectionTimeout ||
          e.type == DioExceptionType.receiveTimeout) {
        throw ApiException.network();
      }
      throw ApiException(
        status: e.response?.statusCode ?? 0,
        message: e.message ?? 'Request failed',
      );
    }
  }
}
